## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#9fe0736d-b05f-4af9-ac39-2af4ce2f1455) |
| mint-fs.sh | :x: [more...](#9185480b-79e5-4888-b655-eb58655882ca) |
| mint-zoned.sh | :x: [more...](#5f622452-5791-45d5-867f-d7395450e6d0) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#6ebbfc0f-87b9-4f48-a1ac-60aaeb32988a) |
<a name="9fe0736d-b05f-4af9-ac39-2af4ce2f1455"></a>
### 12210-97db6e7/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:30206
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b1b8caeae993:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 2 minutes and 18 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 54 seconds
(8/15) Running minio-dotnet tests ... done in 55 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 40 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 13 seconds
(11/15) Running minio-js tests ... FAILED in 17 seconds
{
  "name": "minio-js",
  "function": "fPutObject(bucketName, objectName, filePath, metaData)",
  "args": "bucketName:minio-js-test-aa79c3e5-ba63-4a24-9fb3-3b4f26a8c185, objectName:datafile-65-MB, filePath:/tmp/datafile-65-MB",
  "duration": 1106,
  "status": "FAIL",
  "error": "Error: read ECONNRESET at TCP.onStreamRead (internal/stream_base_commons.js:209:20)"
}
(11/15) Running minio-py tests ... done in 2 minutes and 20 seconds
(12/15) Running s3cmd tests ... done in 17 seconds
(13/15) Running s3select tests ... done in 6 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="5f622452-5791-45d5-867f-d7395450e6d0"></a>
### 12210-97db6e7/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31885
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 0d9a763ccb16:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 1 minutes and 26 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 29 seconds
(8/15) Running minio-dotnet tests ... done in 47 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 24 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 52 seconds
(11/15) Running minio-js tests ... done in 53 seconds
(12/15) Running minio-py tests ... FAILED in 2 minutes and 36 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-081a1368-16a3-4f9c-b1df-42204e69037d",
    "object_name": "c2a9b8ff-710d-411b-b4a0-93682df74eb1"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-081a1368-16a3-4f9c-b1df-42204e69037d, request_id: 167BBEDD1264EE27, host_id: cfc6aab3-0282-4412-a236-e2799df2f4b2, bucket_name: minio-py-test-081a1368-16a3-4f9c-b1df-42204e69037d",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1640, in test_thread_safe\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-081a1368-16a3-4f9c-b1df-42204e69037d, request_id: 167BBEDD1264EE27, host_id: cfc6aab3-0282-4412-a236-e2799df2f4b2, bucket_name: minio-py-test-081a1368-16a3-4f9c-b1df-42204e69037d\n",
  "duration": 5855
}
(12/15) Running s3cmd tests ... done in 18 seconds
(13/15) Running s3select tests ... done in 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="6ebbfc0f-87b9-4f48-a1ac-60aaeb32988a"></a>
### 12210-97db6e7/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:31536
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 06a6bc1049f1:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... FAILED in 31 seconds
{
  "name": "awscli",
  "duration": 2561,
  "function": "aws --endpoint-url http://minio-dev8.minio.io:31536 s3api copy-object --bucket awscli-mint-test-bucket-17075 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-17075/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 46 seconds
(7/15) Running minio-dotnet tests ... done in 41 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 13 seconds
{
  "args": {},
  "duration": 507,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 2 minutes and 1 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 140,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-1ckqibo, objectName = minio-java-test-jp085r-copy, resource = /minio-java-test-1ckqibo/minio-java-test-jp085r-copy, requestId = 167BBEA125B68A11, hostId = 7dcbf806-d186-4b84-8827-eec428623f80)\nrequest={method=PUT, url=http://minio-dev8.minio.io:31536/minio-java-test-1ckqibo/minio-java-test-jp085r-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-2s9834q/minio-java-test-jp085r\nHost: minio-dev8.minio.io:31536\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210504T031450Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210504/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 416\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Tue, 04 May 2021 03:14:50 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 167BBEA125B68A11\nX-Xss-Protection: 1; mode=block\nDate: Tue, 04 May 2021 03:14:50 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 48 seconds
(9/15) Running minio-py tests ... done in 2 minutes and 55 seconds
(10/15) Running s3cmd tests ... FAILED in 4 seconds
{
  "name": "s3cmd",
  "duration": "2665",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-11003/s3cmd-test-object-14572 900231b2-dd2a-4b93-bb3d-f6a495a148c6\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 900231b2-dd2a-4b93-bb3d-f6a495a148c6 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 6 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
<a name="9185480b-79e5-4888-b655-eb58655882ca"></a>
### 12210-97db6e7/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:31483
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7a7ca9676b80:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 0 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 40 seconds
(8/15) Running minio-dotnet tests ... done in 36 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 3 seconds
{
  "args": {
    "bucketName": "minio-go-test-lkutr4lo4vyxewim",
    "objectName": "gynouj3ivfepswzxu9klkmfk9vwruj"
  },
  "duration": 12104,
  "error": "read tcp 172.17.0.5:39386->72.28.97.61:31483: read: connection reset by peer",
  "function": "GetObject(bucketName, objectName)",
  "message": "ReadAt failed",
  "name": "minio-go: testSSES3EncryptedGetObjectReadAtFunctional",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 21 seconds
(10/15) Running minio-js tests ... done in 54 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 8 seconds
(12/15) Running s3cmd tests ... done in 13 seconds
(13/15) Running s3select tests ... done in 2 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/admin-handlers-users.go
+++ b/cmd/admin-handlers-users.go
@@ -490,12 +490,6 @@ func (a adminAPIHandlers) AddServiceAccount(w http.ResponseWriter, r *http.Reque
 		return
 	}
 
-	// Disallow creating service accounts by root user.
-	if createReq.TargetUser == globalActiveCred.AccessKey {
-		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrAdminAccountNotEligible), r.URL)
-		return
-	}
-
 	var (
 		targetUser   string
 		targetGroups []string

--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -880,7 +880,7 @@ func (sys *IAMSys) SetTempUser(accessKey string, cred auth.Credentials, policyNa
 		// This mapping is necessary to ensure that valid credentials
 		// have necessary ParentUser present - this is mainly for only
 		// webIdentity based STS tokens.
-		if cred.IsTemp() && cred.ParentUser != "" {
+		if cred.IsTemp() && cred.ParentUser != "" && cred.ParentUser != globalActiveCred.AccessKey {
 			if _, ok := sys.iamUserPolicyMap[cred.ParentUser]; !ok {
 				if err := sys.store.saveMappedPolicy(context.Background(), accessKey, stsUser, false, mp, options{ttl: ttl}); err != nil {
 					sys.store.unlock()
@@ -1114,14 +1114,10 @@ func (sys *IAMSys) NewServiceAccount(ctx context.Context, parentUser string, gro
 	sys.store.lock()
 	defer sys.store.unlock()
 
-	if parentUser == globalActiveCred.AccessKey {
-		return auth.Credentials{}, errIAMActionNotAllowed
-	}
-
 	cr, ok := sys.iamUsersMap[parentUser]
 	if !ok {
-		// For LDAP users we would need this fallback
-		if sys.usersSysType != MinIOUsersSysType {
+		// For LDAP/OpenID users we would need this fallback
+		if sys.usersSysType != MinIOUsersSysType && parentUser != globalActiveCred.ParentUser {
 			_, ok = sys.iamUserPolicyMap[parentUser]
 			if !ok {
 				var found bool
@@ -1479,7 +1475,12 @@ func (sys *IAMSys) GetUser(accessKey string) (cred auth.Credentials, ok bool) {
 		if cred.IsServiceAccount() || cred.IsTemp() {
 			// temporary credentials or service accounts
 			// must have their parent in UsersMap
-			_, ok = sys.iamUserPolicyMap[cred.ParentUser]
+			if cred.ParentUser == globalActiveCred.AccessKey {
+				// parent exists, so allow temporary and service accounts.
+				ok = true
+			} else {
+				_, ok = sys.iamUserPolicyMap[cred.ParentUser]
+			}
 		}
 		// for LDAP service accounts with ParentUser set
 		// we have no way to validate, either because user
@@ -1865,13 +1866,17 @@ func (sys *IAMSys) policyDBGet(name string, isGroup bool) (policies []string, er
 	var u auth.Credentials
 	var ok bool
 	if sys.usersSysType == MinIOUsersSysType {
+		if name == globalActiveCred.AccessKey {
+			return []string{"consoleAdmin"}, nil
+		}
+
 		// When looking for a user's policies, we also check if the user
 		// and the groups they are member of are enabled.
-
 		u, ok = sys.iamUsersMap[name]
 		if !ok {
 			return nil, errNoSuchUser
 		}
+
 		if !u.IsValid() {
 			return nil, nil
 		}

--- a/cmd/sts-handlers.go
+++ b/cmd/sts-handlers.go
@@ -122,24 +122,20 @@ func checkAssumeRoleAuth(ctx context.Context, r *http.Request) (user auth.Creden
 		if APIErrorCode(s3Err) != ErrNone {
 			return user, false, STSErrorCode(s3Err)
 		}
-		var owner bool
-		user, owner, s3Err = getReqAccessKeyV4(r, globalServerRegion, serviceSTS)
+
+		user, _, s3Err = getReqAccessKeyV4(r, globalServerRegion, serviceSTS)
 		if APIErrorCode(s3Err) != ErrNone {
 			return user, false, STSErrorCode(s3Err)
 		}
-		// Root credentials are not allowed to use STS API
-		if owner {
+
+		// Temporary credentials or Service accounts cannot generate further temporary credentials.
+		if user.IsTemp() || user.IsServiceAccount() {
 			return user, true, ErrSTSAccessDenied
 		}
 	}
 
 	// Session tokens are not allowed in STS AssumeRole requests.
 	if getSessionToken(r) != "" {
-		return user, true, ErrSTSAccessDenied
-	}
-
-	// Temporary credentials or Service accounts cannot generate further temporary credentials.
-	if user.IsTemp() || user.IsServiceAccount() {
 		return user, true, ErrSTSAccessDenied
 	}
 
@@ -157,6 +153,7 @@ func (sts *stsAPIHandlers) AssumeRole(w http.ResponseWriter, r *http.Request) {
 		writeSTSErrorResponse(ctx, w, isErrCodeSTS, stsErr, nil)
 		return
 	}
+
 	if err := r.ParseForm(); err != nil {
 		writeSTSErrorResponse(ctx, w, true, ErrSTSInvalidParameterValue, err)
 		return

